src/librustc_parse/parser/item.rs

```diff
@@ -1740,6 +1742,26 @@ impl<'a> Parser<'a> {
         ).emit();
     }
 
+    /// Checks if current token is keyword that is currently being parsed and reports an error if it is.
+    fn report_nested_keyword(&mut self, kw: Symbol) -> PResult<'a, ()> {
+        if self.token.is_keyword(kw) {
+            let mut err = self.struct_span_err(
+                self.token.span,
+                &format!("`{}` definitions cannot be nested", kw),
+            );
+
+            err.span_suggestion(
+                self.token.span,
+                &format!("consider creating a new `{}` definition instead of nesting", kw),
+                String::new(),
+                Applicability::MaybeIncorrect,
+            );
+            err.emit();
```

Comment:
You don't need to `return Err(err)` if you use recovery. Once you've done that, you can `.emit()` and `.span_suggestion()` can be chained here.

Reply:
I've decided to return error here to prevent any more errors to appear. Otherwise, user is flooded with 4 different errors and I thought it can be overwhelming.

Should I just stick with `emit`?

Reply:
More errors shouldn't appear if you use `self.parse_item()?;`. ;)